controller/scan/harbor.go

```diff
@@ -0,0 +1,191 @@
+package scan
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/neuvector/neuvector/share"
+	"github.com/neuvector/neuvector/share/httpclient"
+	log "github.com/sirupsen/logrus"
+)
+
+type harbor struct {
+	base
+	client http.Client
+}
+
+type HarborApiProject struct {
+	Id         int     `json:"id"`
+	RegistryId *string `json:"registry_id"`
+}
+
+func (proj *HarborApiProject) IsProxyCacheProject() bool {
+	return proj.RegistryId != nil
+}
+
+type HarborApiRepository struct {
+	FullName  string `json:"name"`
+	ProjectId int    `json:"project_id"`
+}
+
+func newHarbor(baseDriver base) *harbor {
+	client := &http.Client{
+		Timeout: gitTimeout,
+	}
+
+	t, err := httpclient.GetTransport(baseDriver.proxy)
+	if err != nil {
+		log.WithError(err).Warn("failed to get transport")
+	} else {
+		client.Transport = t
+	}
+
+	return &harbor{
+		base:   baseDriver,
+		client: *client,
+	}
+}
+
+// TODO: the splits in projectName() and repoName() are duplicated, perform only once
+func (harborRepo *HarborApiRepository) projectName() string {
+	splitRepoName := strings.Split(harborRepo.FullName, "/")
+	if len(splitRepoName) < 2 {
+		// a harbor repo should always have the project name
+		// prefixed which means this should never happen
+		// but handling for the sake of robustness
+		return ""
+	}
+	// A Harbor project is not allowed to have forward-slash
+	// characters, so the first one we run into in a repository
+	// name delimits the harbor project name that repo belongs to
+	return splitRepoName[0]
+}
+
+func (harborRepo *HarborApiRepository) repoName() string {
+	splitRepoName := strings.Split(harborRepo.FullName, "/")
+	if len(splitRepoName) < 2 {
+		return harborRepo.FullName
+	}
+	return strings.Join(splitRepoName[1:], "/")
+}
+
+func (h *harbor) GetAllImages() (map[share.CLUSImage][]string, error) {
+	repositories, err := h.getAllRepositories()
+	if err != nil {
+		return nil, fmt.Errorf("could not get repositories from harbor api: %s", err.Error())
+	}
+
+	images := map[share.CLUSImage][]string{}
+	for _, repository := range repositories {
+		image := share.CLUSImage{
+			Repo: repository.FullName,
+		}
+		tags, err := h.getTagsForRepository(repository)
+		if err != nil {
+			return nil, fmt.Errorf("could not get tags for repository %s from harbor api: %s", repository.FullName, err.Error())
+		}
+		images[image] = tags
+	}
+
+	return images, nil
+}
+
+// TODO: deal with large registries, implement pagination/chunked responses?
+func (h *harbor) getAllRepositories() ([]HarborApiRepository, error) {
+	req, err := http.NewRequest("GET", fmt.Sprintf("%s/api/v2.0/repositories", h.base.regURL), nil)
```

Comment:
Consider using [url.JoinPath](https://pkg.go.dev/net/url#JoinPath)?  It's generally a safer choice because it cleans the URL and removes elements that can be used for path traversal.